tests/qa_automation/qa_run.pm

```diff
@@ -77,144 +74,76 @@ sub prepare_repos {
         die "No QA_HEAD_REPO specified!";
     }
-    assert_script_run "zypper --no-gpg-check -n ar -f '$qa_head_repo' qa-ibs";
+    zypper_call("--no-gpg-check ar -f '$qa_head_repo' qa-ibs");
     if ($qa_web_repo) {
-        assert_script_run "zypper --no-gpg-check -n ar -f '$qa_web_repo' qa-web";
-    }
-    else {
-        script_run("echo 'info: No QA_WEB_REPO configured in this testsuit.'");
+        zypper_call("--no-gpg-check ar -f '$qa_web_repo' qa-web");
     }
     # sometimes updates.suse.com is busy, so we need to wait for possiblye retries
-    assert_script_run "zypper --gpg-auto-import-keys ref", 600;
+    zypper_call("--gpg-auto-import-keys ref");
     zypper_call("in qa_testset_automation qa_tools");
 }
 
 # Create qaset/config file, reset qaset, and start testrun
 sub start_testrun {
     my $self = shift;
-    $self->create_qaset_config();
+    $self->qaset_config();
     assert_script_run("/usr/share/qa/qaset/qaset reset");
-    my $testsuite = $self->test_suite();
-    assert_script_run("/usr/share/qa/qaset/run/$testsuite-run.openqa");
-}
-
-# Safely run shell commands and get output
-sub qa_script_output {
-    my ($self, $cmd, $timeout) = @_;
-    $timeout //= 90;
-    my $random      = int(rand(999999));
-    my $output_file = "/tmp/SCRIPT_OUTPUT_$random";
-    # Run cmd and save output to file
-    script_run(qq{$cmd 2>&1 | tee $output_file}, $timeout);
-    # Write output to serial console
-    my $output_cmd = "sleep 0.1; echo SCRIPT_BEGIN_$random >> /dev/$serialdev; ";
-    $output_cmd .= "cat $output_file >> /dev/$serialdev; ";
-    $output_cmd .= "echo SCRIPT_END_$random >> /dev/$serialdev";
-    script_run($output_cmd, 0);
-    # Get output from serial console
-    my $output = wait_serial("SCRIPT_END_$random", 30);
-    $output =~ s/.*?SCRIPT_BEGIN_$random//sg;
-    $output =~ s/SCRIPT_END_$random.*//sg;
-    $output =~ s/^\s+|\s+$//g;
-    # Remove output file
-    script_run("rm -f $output_file", 10);
-    return $output;
+    assert_script_run("/usr/share/qa/qaset/run/kernel-all-run.openqa");
 }
 
-# Check whether DONE file exists every $interval secs in the background
+# Check whether DONE file exists every $sleep secs in the background
 sub wait_testrun {
-    my ($self, $interval) = @_;
-    $interval //= 30;
-    my $done_file = '/var/log/qaset/control/DONE';
-    my $pattern   = "TESTRUN_FINISHED-" . int(rand(999999));
-    script_run("while [[ ! -f $done_file ]]; do sleep $interval; done; echo $pattern >> /dev/$serialdev", 0);
+    my $self    = shift;
+    my %args    = @_;
+    my $timeout = $args{timeout} || 180 * 60;
+    my $sleep   = $args{sleep} || 30;
+    my $fdone   = '/var/log/qaset/control/DONE';
+    my $pattern = "TESTRUN_FINISHED";
+    my $code    = int(rand(999999));
+    my $redir   = (is_serial_terminal) ? "" : " >> /dev/$serialdev";
+    my $cmd     = "code=$code; while [[ ! -f $fdone ]]; do sleep $sleep; done; echo \"$pattern-\$code\" $redir";
+    type_string("bash -c '$cmd' &\n");
```

Comment:
Maybe you should use `script_output` instead of `type_string` and `wait_serial`. `script_output` uploads the script and calls bash for you (see os-autoinst/testapi.pm). It might not work how you want though.

This looks like a big improvement in general.